File: contracts/evidence.py
"""evidence-pack/1 — 판단 계층 입력 계약 (SSOT).

detection 서비스가 조립하고 judgment 서비스가 소비한다.
서브에이전트/LLM은 대화 이력에 접근할 수 없으므로,
판단에 필요한 모든 맥락은 이 팩에 물질화되어야 한다.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from contracts.snapshot import FiredIndicator, Phase


class MarketSnapshot(BaseModel):
    """평가 틱 시점의 주요 시세 요약 (as_of 기준, PIT 보장)."""

    as_of: datetime
    kospi: float
    kospi_chg_1d_pct: float
    kospi_drawdown_60d_pct: float
    usdkrw: float
    usdkrw_chg_1d_pct: float
    vix: float | None = None
    vkospi: float | None = None
    hy_oas: float | None = None
    spx_chg_1d_pct: float | None = None


class NewsCluster(BaseModel):
    """P2부터 채워짐. P1에서는 빈 리스트."""

    topic: str
    article_count: int
    novelty_z: float | None = None
    representative_headlines: list[str] = Field(max_length=5)


class MacroEvent(BaseModel):
    date: str  # YYYY-MM-DD
    name: str  # 예: "FOMC", "금통위", "미 CPI"


class AnalogueRef(BaseModel):
    """P3 analogue library 조회 결과. P1에서는 빈 리스트."""

    event_id: str
    similarity: float
    outcome_summary: str  # 예: "KOSPI 5d -6.2%, 20d -3.1%, 60d +4.0%"


class EvidencePack(BaseModel):
    schema_id: Literal["evidence-pack/1"] = Field("evidence-pack/1", alias="schema")
    built_at: datetime
    phase: Phase
    prev_phase: Phase
    composite_score: float
    fired_indicators: list[FiredIndicator]
    market: MarketSnapshot
    news_clusters: list[NewsCluster] = Field(default_factory=list)
    macro_calendar_7d: list[MacroEvent] = Field(default_factory=list)
    analogues: list[AnalogueRef] = Field(default_factory=list)

    model_config = {"populate_by_name": True}

File: mobile/engine/src/main/kotlin/com/branchconsole/engine/contracts/Evidence.kt
package com.branchconsole.engine.contracts

import kotlinx.datetime.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Kotlin mirror of contracts/evidence.py (evidence-pack/1, pydantic SSOT).
 *
 * See Snapshot.kt's file header for the shared wire regulations (M1_PLAN_B.md §6.3).
 * Reuses [Phase] and [FiredIndicator] from Snapshot.kt — evidence.py itself imports both
 * from contracts.snapshot, so this mirrors the same cross-file relationship.
 */

private const val HEADLINES_MAX = 5
private const val EVIDENCE_PACK_SCHEMA_ID = "evidence-pack/1"

@Serializable
data class MarketSnapshot(
    @SerialName("as_of") val asOf: Instant,
    val kospi: Double,
    @SerialName("kospi_chg_1d_pct") val kospiChg1dPct: Double,
    @SerialName("kospi_drawdown_60d_pct") val kospiDrawdown60dPct: Double,
    val usdkrw: Double,
    @SerialName("usdkrw_chg_1d_pct") val usdkrwChg1dPct: Double,
    val vix: Double? = null,
    val vkospi: Double? = null,
    @SerialName("hy_oas") val hyOas: Double? = null,
    @SerialName("spx_chg_1d_pct") val spxChg1dPct: Double? = null,
)

@Serializable
data class NewsCluster(
    val topic: String,
    @SerialName("article_count") val articleCount: Int,
    @SerialName("novelty_z") val noveltyZ: Double? = null,
    @SerialName("representative_headlines") val representativeHeadlines: List<String>,
) {
    init {
        require(representativeHeadlines.size <= HEADLINES_MAX) {
            "representative_headlines must have <= $HEADLINES_MAX elements, " +
                "was ${representativeHeadlines.size}"
        }
    }
}

@Serializable
data class MacroEvent(
    val date: String,
    val name: String,
)

@Serializable
data class AnalogueRef(
    @SerialName("event_id") val eventId: String,
    val similarity: Double,
    @SerialName("outcome_summary") val outcomeSummary: String,
)

@Serializable
data class EvidencePack(
    @SerialName("schema") val schema: String = EVIDENCE_PACK_SCHEMA_ID,
    @SerialName("built_at") val builtAt: Instant,
    val phase: Phase,
    @SerialName("prev_phase") val prevPhase: Phase,
    @SerialName("composite_score") val compositeScore: Double,
    @SerialName("fired_indicators") val firedIndicators: List<FiredIndicator>,
    val market: MarketSnapshot,
    @SerialName("news_clusters") val newsClusters: List<NewsCluster> = emptyList(),
    @SerialName("macro_calendar_7d") val macroCalendar7d: List<MacroEvent> = emptyList(),
    val analogues: List<AnalogueRef> = emptyList(),
) {
    init {
        require(schema == EVIDENCE_PACK_SCHEMA_ID) {
            "schema must be '$EVIDENCE_PACK_SCHEMA_ID', was '$schema'"
        }
    }
}
